### PR TITLE
Adds Navio2 RC Input & PWM Output

### PR DESCRIPTION
**Pull request by mayanez**
We've hardcoded some values we think are configurable such as number of channels and outputs. Not sure what the correct way to do that is. 

Other than that, we've tested using our RC transmitter and verified that the servos operate correctly on our drone.

### CODE REVIEW

**Comment by mhkabir** [comment]
Nicely done :) 

Would've been great if we could've just used a modified version of our own PX4IO drivers, but that may come later.

**Comment by mayanez** [comment]
Yeah, I talked with Lorenz about that. I told him I would first do it using their sysfs drivers to get the drone flight ready. Once I can verify everything works end to end, I'll implement the PX4IO driver using the DriverFramework.

**Comment by staroselskii** [comment]
@mayanez @mhkabir 
The work you've done is awesome. 
There's one thing that concerns me, though. I don't understand why bother putting effort into two implementations of the same thing? We would all be better off concentrating on the same implementation, wouldn't we?

**Comment by mayanez** [comment]
@staroselskii 
We weren't aware anyone else was adding support for the PX4 stack at the moment and are trying to add support as quickly as possible for a research project we are doing.  

As a side note, when using `actuator_control_0` instead of `actuator_control_3`, which is the manual bypass group, and we have QGroundcontrol connected via radio telemetry we get stuttering on the motors.Withouth QGroundcontrol in the loop it works fine. Does anyone have any idea why this might be?

**Comment by mhkabir** [comment]
I've seen the telemetry radios creating interference before.
As such, the radios don't have sufficient bandwidth for  proper manual control, so you might be running into that as well. Do you see any  "manual control lost" warnings in QGC?

**Comment by mayanez** [comment]
The strange part is if I use another GC such as MavProxy.py I have no issues. I'm not currently using the telemetry radio for manual control, instead a FrSky receiver. It seems somehow QGC might be overloading the control loop somehow.

**Comment by LorenzMeier** [comment]
Do you have a joystick connected in QGroundControl? That would make it send joystick commands.

**Comment by LorenzMeier** [comment]
@mayanez The build was rejected because the code style was not a match. Can you please run Tools/check_code_style_all.sh and run the formatting fix script as advised there?

**Comment by mayanez** [comment]
I've updated the code style and verified the issue with the QGroundControl.

@LorenzMeier You were right, I had the virtual joystick setting enabled which was causing issues. Just curious to understand how exactly that can be prevented. Does it happen because of task scheduling and polling rates?

**Comment by mhkabir** [comment]
The joystick commands and the RC commands were interfering, simple.

**Comment by LorenzMeier** [comment]
@mayanez I would like to rename / move the two apps slightly:

  * both in src/drivers
  * both named either navio_pwm_out and navio_rc_in or sysfs_pwm_out and sysfs_rc_in

**Comment by LorenzMeier** [comment]
@mayanez And a config in posix-configs would be great which includes a script starting all drivers you need for NAVIO operation (unless that's already there and I missed it).

**Comment by LorenzMeier** [comment]
@mayanez And last but not least: It would be fantastic to get a PR for the docs with Navio 2 instructions:
http://dev.px4.io/hardware-pi2.html
pages / repo are here:
https://github.com/PX4/Devguide

File: posix-configs/rpi2/mainapp.config [hunk 1]
```diff
@@ -19,3 +19,5 @@ sleep 1
 mavlink stream -u 14556 -s HIGHRES_IMU -r 50
 mavlink stream -u 14556 -s ATTITUDE -r 50
 mavlink boot_complete
+rcinput start
```
Comment (LorenzMeier) [inline posix-configs/rpi2/mainapp.config]: This is named differently from the actual driver

File: cmake/configs/posix_rpi2_release.cmake [hunk 2]
```diff
@@ -72,9 +72,12 @@ set(config_module_list
 	modules/navigator
 	modules/mavlink
 
+	modules/rcinput
```
Comment (LorenzMeier) [inline cmake/configs/posix_rpi2_release.cmake]: This is named differently from the actual driver.

**Comment by LorenzMeier** [comment]
I'm really sorry to bother you still with feedback, but the PR looks like you haven't flown this, as the app would not be built / started.

I would appreciate if you could test-fly this very branch and make sure it all works. That's the testing level we usually require for the final commit in a PR, as its the only way to make sure that others from the development community can base off your work.

**Comment by HidenoriKobayashi** [comment]
@LorenzMeier 
For the last push, it is my fault. I apologize for that. We have created a new build target for navio2 specifically because we thought rpi2 was too generic for what we are doing and forgot to (1) modify the cmake file for the default RPi2 and (2) add config for the navio that we are currently using. With this new configuration file, we have verified that the servos are under the control of our RC transmitter. I will push these changes today.

Yes I do understand the importance of keeping the quality of code base to a certain good level.

For the benefit of the community, I think we need a page that describes what modules worked under what conditions by how and make it a requirement to update the page when necessary upon creating a pull request. We can make a starting point for our work on navio2. Also, I feel we need a better testing framework and criteria for drivers, because then people can start building their work on those drivers before it is actually verified by test flight. My apologies if there is already such a thing and I missed it.

**Comment by LorenzMeier** [comment]
@HidenoriKobayashi Your feedback is spot on and appreciated. I'm all for even tighter testing and docs and it would be fantastic if you could help push that.

**Comment by LorenzMeier** [comment]
Rebased and applied, thanks!

**Comment by cudnn** [comment]
navio 2 board use diffrent  spi between stm32 and imu。i want to use the same spi 。

**Pull request closed without merging**